File: examples/cpp/Misc/PipelineDebugging/node_pipeline_events.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

#include "depthai/depthai.hpp"

int main() {
    dai::Pipeline pipeline;
    pipeline.enablePipelineDebugging();

    auto monoLeft = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B);
    auto monoRight = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C);

    auto stereo = pipeline.create<dai::node::StereoDepth>();

    auto monoLeftOut = monoLeft->requestFullResolutionOutput();
    auto monoRightOut = monoRight->requestFullResolutionOutput();

    monoLeftOut->link(stereo->left);
    monoRightOut->link(stereo->right);

    stereo->setRectification(true);
    stereo->setExtendedDisparity(true);
    stereo->setLeftRightCheck(true);

    auto depthQueue = stereo->depth.createOutputQueue();
    auto monoLeftEventQueue = monoLeft->pipelineEventOutput.createOutputQueue(1, false);

    pipeline.start();
    while(pipeline.isRunning()) {
        auto depth = depthQueue->get<dai::ImgFrame>();
        auto latestNodeEvent = monoLeftEventQueue->tryGet<dai::PipelineEvent>();
        cv::imshow("depth", dai::utility::colorizeDepthFrame(*depth, 500.0f, 12000.0f, cv::COLORMAP_JET, true).getCvFrame());

        std::cout << "Latest event from MonoLeft camera node: " << (latestNodeEvent ? latestNodeEvent->str() : "No event");

        int key = cv::waitKey(1);
        if(key == 'q') {
            break;
        }
    }

    pipeline.stop();

    return 0;
}

File: examples/python/Misc/PipelineDebugging/node_pipeline_events.py
import time

import depthai as dai
import cv2

with dai.Pipeline() as pipeline:
    # Pipeline debugging is disabled by default.
    # You can also enable it by setting the DEPTHAI_PIPELINE_DEBUGGING environment variable to '1'
    pipeline.enablePipelineDebugging(True)

    monoLeft = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B)
    monoRight = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C)
    stereo = pipeline.create(dai.node.StereoDepth)

    # Linking
    monoLeftOut = monoLeft.requestFullResolutionOutput()
    monoRightOut = monoRight.requestFullResolutionOutput()
    monoLeftOut.link(stereo.left)
    monoRightOut.link(stereo.right)

    stereo.setRectification(True)
    stereo.setExtendedDisparity(True)
    stereo.setLeftRightCheck(True)

    depthQueue = stereo.depth.createOutputQueue()
    monoLeftEventQueue = monoLeft.pipelineEventOutput.createOutputQueue() # Supported on core and rvc4 only

    pipeline.start()
    while pipeline.isRunning():
        depth = depthQueue.get()
        latestEvent = monoLeftEventQueue.tryGet()
        assert isinstance(depth, dai.ImgFrame)
        colorizedDepth = dai.colorizeDepthFrame(depth, 500, 12000, cv2.COLORMAP_JET, useLog=True).getCvFrame()
        cv2.imshow("depth", colorizedDepth)
        print(f"Latest event from MonoLeft camera node: {latestEvent if latestEvent is not None else 'No event'}")
        key = cv2.waitKey(1)
        if key == ord('q'):
            pipeline.stop()
            break
